python/ray/experimental/data_loader/shuffle.py

```diff
@@ -0,0 +1,533 @@
+import itertools
+import time
+import timeit
+import threading
+from typing import Callable, List, Iterable, Union
+
+import pandas as pd
+import numpy as np
+from smart_open import open
+
+import ray
+from ray.experimental.data_loader.stats import (
+    TrialStatsCollector, TrialStatsCollectorFromMemory, collect_store_stats,
+    TrialStatsFromMemory)
+
+
+#
+# In-memory shuffling, loads data from disk once per epoch.
+#
+
+
+def shuffle_from_memory_with_stats(
+        filenames: List[str],
+        batch_consumer: Callable[[int, Iterable[pd.DataFrame]], None],
+        num_epochs: int,
+        num_rounds: int,
+        num_reducers: int,
+        num_trainers: int,
+        max_concurrent_epochs: int,
+        utilization_sample_period: float) -> (TrialStatsFromMemory, List):
+    """
+    Shuffle the provided dataset every epoch, using round-based shuffling that
+    caches the input dataset in cluster memory at the beginning of each epoch.
+    """
+    stats = None
+    store_stats = []
+    done_event = threading.Event()
+    store_stats_collector_thread = threading.Thread(
+        target=collect_store_stats,
+        args=(store_stats, done_event, utilization_sample_period))
+    try:
+        store_stats_collector_thread.start()
+
+        print(
+            f"Doing {num_epochs} epochs each with {num_rounds} shuffle "
+            "rounds.")
+
+        stats = shuffle_from_memory(
+            filenames,
+            batch_consumer,
+            num_epochs,
+            num_rounds,
+            num_reducers,
+            num_trainers,
+            max_concurrent_epochs,
+            collect_stats=True)
+    finally:
+        # Signal store stats collector thread that we're done, join the
+        # thread.
+        done_event.set()
+        store_stats_collector_thread.join()
+
+    return stats, store_stats
+
+
+def shuffle_from_memory_no_stats(
+        filenames: List[str],
+        batch_consumer: Callable[[int, Iterable[pd.DataFrame]], None],
+        num_epochs: int,
+        num_rounds: int,
+        num_reducers: int,
+        num_trainers: int,
+        max_concurrent_epochs: int,
+        utilization_sample_period: float) -> (float, None):
+    """
+    Shuffle the provided dataset every epoch, using round-based shuffling that
+    caches the input dataset in cluster memory at the beginning of each epoch.
+    """
+    print(f"Doing {num_epochs} epochs each with {num_rounds} shuffle rounds.")
+    duration = shuffle_from_memory(
+        filenames,
+        batch_consumer,
+        num_epochs,
+        num_rounds,
+        num_reducers,
+        num_trainers,
+        max_concurrent_epochs,
+        collect_stats=False)
+    return duration, None
+
+
+def shuffle_from_memory(
+        filenames: List[str],
+        batch_consumer: Callable[[int, Iterable[pd.DataFrame]], None],
+        num_epochs: int,
+        num_rounds: int,
+        num_reducers: int,
```

Comment:
We probably want to make this num_reducers_per_trainer, so that we can colocate trainers and reduce tasks. And we also want the option of more reducer tasks than trainers so we can take advantage of parallelism for the reduce stage.

We should also allow passing in a placement group for the reducers so that these can be colocated with the consumers specified by the caller.